services-core/src/solution_submission/retry.rs

```diff
@@ -41,156 +31,60 @@ pub enum RetryResult<TransactionResult, CancellationResult> {
-        Self::with_sleep_and_now(
-            gas_price_estimating,
-            util::AsyncSleep {},
-            util::default_now(),
-        )
-    }
-
-    pub fn with_sleep_and_now(
-        gas_price_estimating: Arc<dyn GasPriceEstimating>,
-        async_sleep: impl AsyncSleeping,
-        now: impl Now,
-    ) -> Self {
-        Self {
-            gas_price_estimating,
-            async_sleep: Box::new(async_sleep),
-            now: Box::new(now),
-        }
-    }
-}
-
-#[async_trait::async_trait]
-impl RetryTransactionSending for RetryWithGasPriceIncrease {
-    async fn retry<TransactionSender, CancellationSender>(
-        &self,
-        transaction_sender: TransactionSender,
-        cancel_after: BoxFuture<'_, CancellationSender>,
-        target_confirm_time: Instant,
-        gas_price_cap: f64,
-    ) -> Option<RetryResult<TransactionSender::Output, CancellationSender::Output>>
-    where
-        TransactionSender: TransactionSending,
-        CancellationSender: TransactionSending,
-    {
-        let gas_price_stream = gas_price_increase::enforce_minimum_increase_and_cap(
-            gas_price_cap,
-            self.gas_price_stream(target_confirm_time),
-        );
-        Self::retry(transaction_sender, cancel_after, gas_price_stream).await
-    }
-}
-
-impl RetryWithGasPriceIncrease {
-    async fn gas_price(&self, target_confirm_time: Instant) -> Result<f64> {
-        let time_remaining = target_confirm_time.saturating_duration_since(self.now.instant_now());
-        // TODO: Use a more accurate gas limit once the gas estimators take that into account.
-        self.gas_price_estimating
-            .estimate_with_limits(SOLUTION_SUBMISSION_GAS_LIMIT as f64, time_remaining)
-            .await
-    }
-
-    // Yields the current gas price immediately and then every refresh interval. Skips errors.
-    fn gas_price_stream(&self, target_confirm_time: Instant) -> impl FusedStream<Item = f64> + '_ {
-        stream::unfold(true, move |first_call| async move {
-            if !first_call {
-                self.async_sleep.sleep(GAS_PRICE_REFRESH_INTERVAL).await;
-            }
-            return Some((self.gas_price(target_confirm_time).await, false));
-        })
-        .filter_map(|gas_price_result| async move {
-            match gas_price_result {
-                Ok(gas_price) => {
-                    log::debug!("estimated gas price {}", gas_price);
-                    Some(gas_price)
-                }
-                Err(err) => {
-                    log::error!("gas price estimation failed: {:?}", err);
-                    None
-                }
+/// Send a transaction whenever the next gas price is ready.
+/// When cancel_after is ready the transaction will be cancelled by sending a noop transaction
+/// at a higher gas price.
+/// Returns None if no transaction has been submitted.
+/// Panics if the stream ends.
+pub async fn retry<TransactionSender, CancellationSender>(
```

Comment:
I removed the trait and made this a standalone function because it is simpler and makes more sense from a standalone crate perspective.

Reply:
I also think this makes more sense as a function.